lib/Dialect/SPIRV/SPIRVOps.cpp

```diff
@@ -199,6 +213,136 @@ static void printNoIOOp(Operation *op, OpAsmPrinter *printer) {
   printer->printOptionalAttrDict(op->getAttrs());
 }
 
+//===----------------------------------------------------------------------===//
+// spv.AccessChainOp
+//===----------------------------------------------------------------------===//
+
+static Type extractPtrTypeFromPtrToCompositeType(ArrayRef<Value *> indices,
+                                                 Type type, Location baseLoc) {
+  if (!indices.size()) {
+    return mlir::emitError(baseLoc, "'spv.AccessChain' op expected at least "
+                                    "one index for spv.AccessChain"),
+           nullptr;
+  }
+
+  auto ptrType = type.dyn_cast<spirv::PointerType>();
+  if (!ptrType) {
+    return mlir::emitError(baseLoc, "'spv.AccessChain' op expected a pointer "
+                                    "to composite type, but provided ")
+               << type,
+           nullptr;
+  }
+
+  auto resultType = ptrType.getPointeeType();
+  auto resultStorageClass = ptrType.getStorageClass();
+  int32_t index = 0;
+
+  for (auto indexSSA : indices) {
+    auto cType = resultType.dyn_cast<spirv::CompositeType>();
+    if (!cType) {
+      return mlir::emitError(
+                 baseLoc,
+                 "'spv.AccessChain' op cannot extract from non-composite type ")
+                 << resultType << " with index " << index,
+             nullptr;
+    }
+    index = 0;
+    if (resultType.isa<spirv::StructType>()) {
+      Operation *op = indexSSA->getDefiningOp();
+      if (!op) {
+        return mlir::emitError(baseLoc, "'spv.AccessChain' op index must be an "
+                                        "integer spv.constant to access "
+                                        "element of spv.struct"),
+               nullptr;
+      }
+
+      if (!extractValueFromConstOp(op, index)) {
```

Comment:
Can you put a TODO here to say that this should be relaxed to allow integer literals of other bitwidths?

Reply:
Updated.